pkg/compactor/compactor.go

```diff
@@ -76,19 +76,20 @@ type BlocksCompactorFactory func(
 
 // Config holds the MultitenantCompactor config.
 type Config struct {
-	BlockRanges           mimir_tsdb.DurationList `yaml:"block_ranges" category:"advanced"`
-	BlockSyncConcurrency  int                     `yaml:"block_sync_concurrency" category:"advanced"`
-	MetaSyncConcurrency   int                     `yaml:"meta_sync_concurrency" category:"advanced"`
-	ConsistencyDelay      time.Duration           `yaml:"consistency_delay" category:"advanced"`
-	DataDir               string                  `yaml:"data_dir"`
-	CompactionInterval    time.Duration           `yaml:"compaction_interval" category:"advanced"`
-	CompactionRetries     int                     `yaml:"compaction_retries" category:"advanced"`
-	CompactionConcurrency int                     `yaml:"compaction_concurrency" category:"advanced"`
-	CleanupInterval       time.Duration           `yaml:"cleanup_interval" category:"advanced"`
-	CleanupConcurrency    int                     `yaml:"cleanup_concurrency" category:"advanced"`
-	DeletionDelay         time.Duration           `yaml:"deletion_delay" category:"advanced"`
-	TenantCleanupDelay    time.Duration           `yaml:"tenant_cleanup_delay" category:"advanced"`
-	MaxCompactionTime     time.Duration           `yaml:"max_compaction_time" category:"advanced"`
+	BlockRanges                mimir_tsdb.DurationList `yaml:"block_ranges" category:"advanced"`
+	BlockSyncConcurrency       int                     `yaml:"block_sync_concurrency" category:"advanced"`
+	MetaSyncConcurrency        int                     `yaml:"meta_sync_concurrency" category:"advanced"`
+	ConsistencyDelay           time.Duration           `yaml:"consistency_delay" category:"advanced"`
+	DataDir                    string                  `yaml:"data_dir"`
+	CompactionInterval         time.Duration           `yaml:"compaction_interval" category:"advanced"`
+	CompactionRetries          int                     `yaml:"compaction_retries" category:"advanced"`
+	CompactionConcurrency      int                     `yaml:"compaction_concurrency" category:"advanced"`
+	CleanupInterval            time.Duration           `yaml:"cleanup_interval" category:"advanced"`
+	CleanupConcurrency         int                     `yaml:"cleanup_concurrency" category:"advanced"`
+	DeletionDelay              time.Duration           `yaml:"deletion_delay" category:"advanced"`
+	TenantCleanupDelay         time.Duration           `yaml:"tenant_cleanup_delay" category:"advanced"`
+	MaxCompactionTime          time.Duration           `yaml:"max_compaction_time" category:"advanced"`
+	EnableBackgroundValidation bool                    `yaml:"-"` // used for unit tests only
```

Comment:
@pstibrany I suggested to Vernon that we disable it by default, in order to avoid unit test authors having to be aware of this option and disabling it explicitly. I reasoned that it would be less hassle to force production code to enable it, rather than the opposite (disabling in test code). If test authors have to disable the option explicitly, I foresee it being a hidden gotcha.

I'm open to other opinions though. If it should be enabled by default, should we ensure that tests have common compactor construction code so test authors don't have to remember disabling background validation?

Reply:
> in order to avoid unit test authors having to be aware of this option and disabling it explicitly.

The only only place where this flag will be used is "complete block upload" handler. I don't think we need more than single unit test this handler.

Validation should definitely be enabled by default. If test code needs to disable it, it can set the flag.

Reply:
@pstibrany if it's only going to affect a single test, I agree it won't be an issue. I got the impression from @aldernero though that multiple tests needed to set the flag?

Reply:
I think there are some tests missing in this PR, but they don't need to call the handler code. (Test can focus on `completeBlockUpload` or validation method directly)